### PR TITLE
feat!: Migrate firebase_analytics to sound null safety

### PR DESCRIPTION
**Pull request by IchordeDionysos**
## Description

This migrates firebase_analytics to non-nullable types

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by IchordeDionysos** [comment]
@Salakar any idea why the iOS build fails? 🤔 

When I run the command locally on my MacBook it works:

```
flutter build ios --no-codesign --simulator --debug --target=./test_driver/firebase_analytics_e2e.dart --dart-define=CI=true
```

File: packages/firebase_analytics/firebase_analytics/test/observer_test.dart
Comment (untp) [inline packages/firebase_analytics/firebase_analytics/test/observer_test.dart:155]: Why `neverEndingFuture`? Using `Future<void>.value()` passes all tests. If it is really needed, `Completer().future` can be used.

Comment (IchordeDionysos) [reply packages/firebase_analytics/firebase_analytics/test/observer_test.dart:155]: But don't all tests already succeed? 🤔 

I've oriented myself here on the @rrousselGit implementation for `firebase_auth` ☺️
https://github.com/FirebaseExtended/flutterfire/blame/master/packages/firebase_auth/firebase_auth/test/mock.dart

Comment (untp) [reply packages/firebase_analytics/firebase_analytics/test/observer_test.dart:155]: Yes, this passes tests but using `Future<void>.value()` or `Completer<void>().future` is much simple.

**Comment by untp** [comment]
I want to make some changes, but I can't edit this PR, so I created a PR to your repo (simpleclub-extended/flutterfire#2). @IchordeDionysos Can you review and merge it? It includes analytics example migration and uses plugin_platform_interface.

File: packages/firebase_analytics/firebase_analytics/test/observer_test.dart
Comment (rrousselGit) [reply packages/firebase_analytics/firebase_analytics/test/observer_test.dart:155]: Ah for `void`, indeed `Future<void>.value` is a better solution. The `neverEdingFuture<void>` are there by mistake you could say.

**Comment by IchordeDionysos** [comment]
@untp thanks for your PR incorporated what I could, unfortunately migrating the example to null-safety is not possible yet integration tests null-safety is not yet in stable.

Also, I did not want to migrate to the platform interfaces to minimize the scope of the PR. :) 

@rrousselGit @Salakar @Ehesp can you review the PR? 😌

File: packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart:31]:
```suggestion
      {required String name, Map<String, Object?>? parameters}) async {
```

Comment (rrousselGit) [reply packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart:31]: Actually no, seems like I'm wrong

File: packages/firebase_analytics/firebase_analytics/test/observer_test.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/test/observer_test.dart:37]: Why is this commented?

File: packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart:31]: It appears that values in `parameters` should be nullable, according to the JS SDK

https://cs.opensource.google/firebase-sdk/firebase-js-sdk/+/master:packages/analytics-interop-types/index.d.ts;l=27?q=analytics%20logEvent&ss=firebase-sdk

So we'd have:

```suggestion
      {required String name, Map<String, Object?>? parameters}) async {
```

File: packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics_web/lib/firebase_analytics_web.dart:23]:
```suggestion
    Map<String, Object?>? parameters,
```

File: packages/firebase_analytics/firebase_analytics_platform_interface/lib/method_channel_firebase_analytics.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics_platform_interface/lib/method_channel_firebase_analytics.dart:19]:
```suggestion
    Map<String, Object?>? parameters,
```

File: packages/firebase_analytics/firebase_analytics/example/ios/Runner.xcodeproj/project.pbxproj
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/example/ios/Runner.xcodeproj/project.pbxproj:152]: Could you remove those changes from the PR? same with the workspacedata file

File: packages/firebase_analytics/firebase_analytics/example/pubspec.yaml
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/example/pubspec.yaml:9]: Is there a reason for this change?

Comment (IchordeDionysos) [reply packages/firebase_analytics/firebase_analytics/example/pubspec.yaml:9]: As e2e is now deprecated and replaced by `integration_test` I hoped that I would gain null-safety, but I then realized that the `intergation_test` null-safety migration did not make this stable release of Flutter ...

File: packages/firebase_analytics/firebase_analytics/example/ios/Runner.xcodeproj/project.pbxproj
Comment (IchordeDionysos) [reply packages/firebase_analytics/firebase_analytics/example/ios/Runner.xcodeproj/project.pbxproj:152]: I hoped that they will fix the iOS build issue which I strangely can't reproduce locally or the build error not showing up in Github actions ...

Comment (IchordeDionysos) [reply packages/firebase_analytics/firebase_analytics/example/ios/Runner.xcodeproj/project.pbxproj:152]: I will remove it again

File: packages/firebase_analytics/firebase_analytics/example/pubspec.yaml
Comment (AlexHartford) [reply packages/firebase_analytics/firebase_analytics/example/pubspec.yaml:9]: Yes, it seems `integration_test` and `flutter_driver` are holding many packages back from null safety. In some cases, developers are foregoing the dependency to enable their users to migrate to null safety.

**Comment by IchordeDionysos** [comment]
@rrousselGit your comments made me investigate this a little bit more 🤔
Initially, I only looked at the firebase-dart package for reference, but now I've looked at all other platforms which parameters they allow nullable.

I've written down how the function headers would look like in Dart for each platform:

[firebase-dart](https://github.com/FirebaseExtended/firebase-dart/blob/master/firebase/lib/src/analytics.dart):
```dart
logEvent(required String name, Map<String, Object?> parameters);
setAnalyticsCollectionEnabled(boolean enabled);
setCurrentScreen(String screenName);
setUserId(String id);
setUserProperties(Map<String, String> properties);
```

[Swift](https://firebase.google.com/docs/reference/swift/firebaseanalytics/api/reference/Classes/Analytics):
```swift
logEvent(required String name, Map<String, Object?>? parameters);
setAnalyticsCollectionEnabled(boolean enabled);
setCurrentScreen(String screenName?, String? screenClassOverride);
setUserId(String? id);
setUserProperties(Map<String, String?> properties);
```

[Kotlin](https://firebase.google.com/docs/reference/kotlin/com/google/firebase/analytics/FirebaseAnalytics):
```swift
logEvent(required String name, Map<String, Object?>? parameters);
setAnalyticsCollectionEnabled(boolean enabled);
setCurrentScreen(String screenName?, String? screenClassOverride);
setUserId(String? id);
setUserProperties(Map<String, String?> properties);
```

So Swift and Kotlin are fully aligned here, but there are some differences that may break the implementation for native users and we should change:

`logEvent`: Make `parameters` values nullable and make `parameters` itself also nullable.
`setCurrentScreen`: Make `screenName` and `screenClassOverride` (not needed for web) nullable.
`setuserId`: Make `id` nullable.
`setUserProperties`: Make `properties` values nullable.

All of those changes are blocked by `firebase-dart`, unfortunately ...

I will open a PR over there and mark this PR as Draft until it's merged and published.

/cc @kevmoo

**Comment by IchordeDionysos** [comment]
@rrousselGit I've adjusted the nullable types in this PR already, but we would have to wait for @kevmoo to publish a new version of `firebase`.

**Comment by IchordeDionysos** [comment]
Switched to the stable channel for all test driver builds as the iOS build of the firebase_analytics example is broken on the dev channel.

See this issue for reference: https://github.com/flutter/flutter/issues/78345

**Comment by IchordeDionysos** [comment]
@rrousselGit I've implemented your suggestions. For it to work properly on web I had to adjust the requirement for `firebase` that you changed 10 days ago in this PR https://github.com/FirebaseExtended/flutterfire/pull/5276.

What do you think of requiring version `^9.0.1` of `firebase`?

**Comment by kevmoo** [comment]
That's a must!

**Comment by sysint64** [comment]
The `id` argument in the `FirebaseAnalytics.setUserId` method should has `String?` type, not `String`.

**Comment by NarHakobyan** [comment]
> The `id` argument in the `FirebaseAnalytics.setUserId` method should has `String?` type, not `String`.

why user id should be null?

**Comment by IchordeDionysos** [comment]
@sysint64 thanks for the hint, somehow forgot to update the types in the main package...

**Comment by IchordeDionysos** [comment]
@Salakar @rrousselGit how should we go forward with this PR to get null-safety for Firebase Analytics? :)

File: .github/workflows/firebase_analytics.yaml
Comment (rrousselGit) [inline .github/workflows/firebase_analytics.yaml:27]: Why this change?

File: packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart:628]: What does `dynamic` mean here?

File: packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart
Comment (rrousselGit) [inline packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart:704]: Same here

**Review by rrousselGit — APPROVED** [review]
LGTM, with some nits

Although I'd like to fix CI issues on master before merging this.

File: .github/workflows/firebase_analytics.yaml
Comment (IchordeDionysos) [reply .github/workflows/firebase_analytics.yaml:27]: @rrousselGit I had to switch to the stable channel because the iOS examples changed some properties that should not be changed, leading to building errors on the dev channel.

On the stable channel, it still works:
https://github.com/flutter/flutter/issues/78345

Should we rather fix the build issues and stick with the dev channel?
OR should we do it in this PR or a separate one?

File: packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart
Comment (IchordeDionysos) [reply packages/firebase_analytics/firebase_analytics/lib/firebase_analytics.dart]: Reverted both :)

File: .github/workflows/firebase_analytics.yaml
Comment (rrousselGit) [reply .github/workflows/firebase_analytics.yaml:27]: The CI has been fixed, so I'll revert this

File: .github/workflows/firebase_analytics.yaml
Comment (rrousselGit) [inline .github/workflows/firebase_analytics.yaml:27]:
```suggestion
        run: ./.github/workflows/scripts/install-flutter.sh dev
```

File: .github/workflows/firebase_analytics.yaml
Comment (rrousselGit) [inline .github/workflows/firebase_analytics.yaml:51]:
```suggestion
        run: ./.github/workflows/scripts/install-flutter.sh dev
```

File: .github/workflows/firebase_analytics.yaml
Comment (rrousselGit) [inline .github/workflows/firebase_analytics.yaml:67]:
```suggestion
        run: ./.github/workflows/scripts/install-flutter.sh dev
```

**Comment by rrousselGit** [comment]
It appears that I do not have the permissions to commit on this PR

Could you sync the branch with master and revert the CI update?

**Comment by IchordeDionysos** [comment]
@rrousselGit I've made the adjustments, but for some reason, the two integration checks for Android/iOS aren't starting successfully and are stuck queueing ...

**Comment by rrousselGit** [comment]
They can take a while to start, because there's a limit on how many checks can be performed at the same time

It appears that they properly started and failed though. I'll retry them

**Comment by rrousselGit** [comment]
Looks like everything passed after a retry.

It's good to go then :smile:

**Comment by NarHakobyan** [comment]
Hi guys, thanks for merging this 👍.

After updating to nullsafety getting error in this part, is it okay to cast it?

image

**Comment by abegehr** [comment]
I'm currently using the git-sources package with `url: https://github.com/simpleclub-extended/flutterfire.git`. Is there already a firebase_analytics null-safety version we can use instead? Otherwise, will this be released in a new version soon?

**Comment by IchordeDionysos** [comment]
@abegehr pre-release version was just published :) 

@rrousselGit what do you suggest with the Object vs dynamic thing that @NarHakobyan brought up.

In general, I try to use Object rather than dynamic. (You know the reasons :D), but in this case it might make the API less clean, as the Flutter API returns dynamic 🤔

I would revert it to the previous behavior. Do you agree?

**Pull request merged**
